From 5fe18c8ab97c3355b04d4de673cc73fe1e6dd4f5 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Tue, 14 Dec 2021 13:15:11 +0200
Subject: [PATCH] [SW-66044] gaudi2: pci: avoid pci core register access

During hot reset handling, if firmware accesses pcie core
register it causes self halt in ARC due to memory error.
This implied that during hot reset pcie core registers
are not accessible temporarily. As a work around this
patch handles hot reset differently for gaudi2. It checks
for hot reset irq after pcie link is up. As ltssm_en is
already set to 0x3 there is no need to reconfigure it.

Change-Id: I2a77f3049ae86731ddc98e2b86e913f8ce3c7a06
---

diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index da461be..985fe04 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -658,7 +658,6 @@
 void pci_dev_specific_hot_rst(void)
 {
 	regmap_writel(0x10, mmPCIE_CORE_HOT_RST);
-	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
 }
 
 void pci_dev_specific_flr(void)
@@ -675,6 +674,23 @@
 			mmCPU_IF_PCIE_SPI_INTR_CLR);
 }
 
+u32 pci_dev_specific_hot_rst_check(void)
+{
+	/* Avoid accessing PCIE core register */
+	return 0;
+}
+
+void pci_dev_specific_post_link_up(void)
+{
+	u32 hot_rst_reg_val;
+
+	hot_rst_reg_val = regmap_readl(mmPCIE_CORE_HOT_RST) & BIT(0);
+	if (hot_rst_reg_val == BIT(0)) {
+		pci_dev_specific_hot_rst();
+		preboot_puts("HOT RST received\n");
+	}
+}
+
 #define MSIX_PBA_OFFSET_REG_VAL		0x400
 
 void set_pci_msix_pba_offset_val(void)
diff --git a/lib/preboot/module_pci.c b/lib/preboot/module_pci.c
index 6cf219c..5c705ff 100644
--- a/lib/preboot/module_pci.c
+++ b/lib/preboot/module_pci.c
@@ -77,6 +77,11 @@
 	}
 }
 
+__weak u32 pci_dev_specific_hot_rst_check(void)
+{
+	return regmap_readl(mmPCIE_CORE_HOT_RST) & BIT(0);
+}
+
 __weak enum cmd_check_res dev_specific_module_pci_check(void)
 {
 	/*
@@ -87,7 +92,7 @@
 		return NOT_FOUND;
 
 	pci_sts.flr_reg_val = regmap_readl(mmPCIE_AUX_FLR_INT);
-	pci_sts.hot_rst_reg_val = regmap_readl(mmPCIE_CORE_HOT_RST) & BIT(0);
+	pci_sts.hot_rst_reg_val = pci_dev_specific_hot_rst_check();
 	pci_sts.rdlh_link_reg_val = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
 	pci_sts.smlh_link_reg_val = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
 
@@ -138,6 +143,8 @@
 	}
 }
 
+__weak void pci_dev_specific_post_link_up(void) {}
+
 void module_pci_run_fast(void)
 {
 	unsigned int rst_src;
@@ -158,6 +165,7 @@
 		preboot_puts_raw("PCIe link is ");
 		if (pci_sts.link_is_up) {
 			preboot_puts("up\n");
+			pci_dev_specific_post_link_up();
 		} else {
 			preboot_puts("down\n");
 			pci_sts.mem_enabled = false;
